### PR TITLE
Enable PPS simulation for Run-3

### PR DESCRIPTION
**Pull request by civanch**
#### PR description:
Flag of LHC transport is set to "True", which means transport to PPS protons scattered to small angles.

This PR is applied only to Run-3 pp WFs, not to PbPb and not to Phase-2.
No effect on CPU and memory is expected. 
 
#### PR validation:
WF 11634.0 + validation by PPS expects

#### if this PR is a backport please specify the original PR and why you need to backport that PR:
no

### CODE REVIEW

**Comment by civanch** [comment]
please test

**Comment by civanch** [comment]
@tvami , @mundim , this PR enables PPS for Run3

**Comment by civanch** [comment]
please test

**Comment by tvami** [comment]
test parameters:
  - workflows = 11634.0

**Comment by civanch** [comment]
+1

observed differences in Run-3 WFs are due to usage of random numbers for PPS. That change simulation histories in central detector

File: SimG4Core/Application/python/g4SimHits_cfi.py [hunk 1]
```diff
@@ -625,10 +625,17 @@
 run2_HCAL_2017.toModify( g4SimHits, HCalSD = dict( TestNumberingScheme = True ) )
 
 ##
-## Disable Castor from Run 3
+## Disable Castor from Run 3, enable PPS
 ##
 from Configuration.Eras.Modifier_run3_common_cff import run3_common
 run3_common.toModify( g4SimHits, CastorSD = dict( useShowerLibrary = False ) ) 
+run3_common.toModify( g4SimHits, LHCTransport = cms.bool(True) )
```
Comment (perrotta) [inline SimG4Core/Application/python/g4SimHits_cfi.py:632]:
```suggestion
run3_common.toModify( g4SimHits, LHCTransport = true )
```

File: SimG4Core/Application/python/g4SimHits_cfi.py [hunk 2]
```diff
@@ -625,10 +625,17 @@
...
+
+##
+## Disable PPS from Run 3 PbPb runs
+##
+from Configuration.Eras.Modifier_pp_on_PbPb_run3_cff import pp_on_PbPb_run3
+pp_on_PbPb_run3.toModify( g4SimHits, LHCTransport = cms.bool(False) )
```
Comment (perrotta) [inline SimG4Core/Application/python/g4SimHits_cfi.py:638]:
```suggestion
pp_on_PbPb_run3.toModify( g4SimHits, LHCTransport = false )
```

File: SimG4Core/Application/python/g4SimHits_cfi.py [hunk 3]
```diff
@@ -652,9 +659,10 @@
...
 
 ##
-## Selection of SD's for Phase2
+## Selection of SD's for Phase2, exclude PPS
 ##
 
 from Configuration.Eras.Modifier_phase2_common_cff import phase2_common
 phase2_common.toModify(g4SimHits,
                        OnlySDs = cms.vstring('ZdcSensitiveDetector', 'TotemT2ScintSensitiveDetector', 'TotemSensitiveDetector', 'RomanPotSensitiveDetector', 'PLTSensitiveDetector', 'MuonSensitiveDetector', 'MtdSensitiveDetector', 'BCM1FSensitiveDetector', 'EcalSensitiveDetector', 'CTPPSSensitiveDetector', 'HGCalSensitiveDetector', 'BSCSensitiveDetector', 'CTPPSDiamondSensitiveDetector', 'FP420SensitiveDetector', 'BHMSensitiveDetector', 'HFNoseSensitiveDetector', 'HGCScintillatorSensitiveDetector', 'CastorSensitiveDetector', 'CaloTrkProcessing', 'HcalSensitiveDetector', 'TkAccumulatingSensitiveDetector') )
+phase2_common.toModify( g4SimHits, LHCTransport = cms.bool(False) )
```
Comment (perrotta) [inline SimG4Core/Application/python/g4SimHits_cfi.py:668]:
```suggestion
phase2_common.toModify( g4SimHits, LHCTransport = false ) 
```

File: SimG4Core/Application/python/g4SimHits_cfi.py [hunk 4]
```diff
@@ -652,9 +659,10 @@
 dd4hep.toModify( g4SimHits, g4GeometryDD4hepSource = True )
 
 ##
-## Selection of SD's for Phase2
+## Selection of SD's for Phase2, exclude PPS
 ##
 
 from Configuration.Eras.Modifier_phase2_common_cff import phase2_common
 phase2_common.toModify(g4SimHits,
                        OnlySDs = cms.vstring('ZdcSensitiveDetector', 'TotemT2ScintSensitiveDetector', 'TotemSensitiveDetector', 'RomanPotSensitiveDetector', 'PLTSensitiveDetector', 'MuonSensitiveDetector', 'MtdSensitiveDetector', 'BCM1FSensitiveDetector', 'EcalSensitiveDetector', 'CTPPSSensitiveDetector', 'HGCalSensitiveDetector', 'BSCSensitiveDetector', 'CTPPSDiamondSensitiveDetector', 'FP420SensitiveDetector', 'BHMSensitiveDetector', 'HFNoseSensitiveDetector', 'HGCScintillatorSensitiveDetector', 'CastorSensitiveDetector', 'CaloTrkProcessing', 'HcalSensitiveDetector', 'TkAccumulatingSensitiveDetector') )
```
Comment (perrotta) [inline SimG4Core/Application/python/g4SimHits_cfi.py:667]:
```suggestion
                       OnlySDs = ['ZdcSensitiveDetector', 'TotemT2ScintSensitiveDetector', 'TotemSensitiveDetector', 'RomanPotSensitiveDetector', 'PLTSensitiveDetector', 'MuonSensitiveDetector', 'MtdSensitiveDetector', 'BCM1FSensitiveDetector', 'EcalSensitiveDetector', 'CTPPSSensitiveDetector', 'HGCalSensitiveDetector', 'BSCSensitiveDetector', 'CTPPSDiamondSensitiveDetector', 'FP420SensitiveDetector', 'BHMSensitiveDetector', 'HFNoseSensitiveDetector', 'HGCScintillatorSensitiveDetector', 'CastorSensitiveDetector', 'CaloTrkProcessing', 'HcalSensitiveDetector', 'TkAccumulatingSensitiveDetector'] )
```

**Review by perrotta — COMMENTED** [review]
A safer syntax is to modify parameters without explicitly showing the cms.sometype.
The same could be also extended to lines 645, 646, 652, 653 in this same config

**Comment by perrotta** [comment]
Just to keep track of it, is there an evaluation somewhere, which can be lonked to the PR description, of the cpu and memory footprint of adding the PPS simulation to run3?

**Comment by civanch** [comment]
@perrotta , I have added a paragraph to description. Can I commit your suggestion for SD?

**Comment by perrotta** [comment]
> @perrotta , I have added a paragraph to description. Can I commit your suggestion for SD?

Thank you @civanch 
Yes, of course I have nothing in contrary if you commit my suggestions for cleaning the python config (even if I don't necessarily ask for it). You are the one who will sign this PR once ready, therefore delays will be kept minimal.

**Comment by civanch** [comment]
please test

File: SimG4Core/Application/python/g4SimHits_cfi.py [hunk 5]
```diff
@@ -625,10 +625,17 @@
...
 ##
 from Configuration.Eras.Modifier_run3_common_cff import run3_common
 run3_common.toModify( g4SimHits, CastorSD = dict( useShowerLibrary = False ) ) 
+run3_common.toModify( g4SimHits, LHCTransport = cms.bool(True) )
+
+##
+## Disable PPS from Run 3 PbPb runs
+##
+from Configuration.Eras.Modifier_pp_on_PbPb_run3_cff import pp_on_PbPb_run3
+pp_on_PbPb_run3.toModify( g4SimHits, LHCTransport = false )
```
Comment (perrotta) [inline SimG4Core/Application/python/g4SimHits_cfi.py:638]: My bad, sorry Vladimir:  "False" should have been written with capital "F"
```suggestion
pp_on_PbPb_run3.toModify( g4SimHits, LHCTransport = False )
```

File: SimG4Core/Application/python/g4SimHits_cfi.py [hunk 6]
```diff
@@ -652,9 +659,10 @@
...
+                       OnlySDs = ['ZdcSensitiveDetector', 'TotemT2ScintSensitiveDetector', 'TotemSensitiveDetector', 'RomanPotSensitiveDetector', 'PLTSensitiveDetector', 'MuonSensitiveDetector', 'MtdSensitiveDetector', 'BCM1FSensitiveDetector', 'EcalSensitiveDetector', 'CTPPSSensitiveDetector', 'HGCalSensitiveDetector', 'BSCSensitiveDetector', 'CTPPSDiamondSensitiveDetector', 'FP420SensitiveDetector', 'BHMSensitiveDetector', 'HFNoseSensitiveDetector', 'HGCScintillatorSensitiveDetector', 'CastorSensitiveDetector', 'CaloTrkProcessing', 'HcalSensitiveDetector', 'TkAccumulatingSensitiveDetector'] )
+phase2_common.toModify( g4SimHits, LHCTransport = false )
```
Comment (perrotta) [inline SimG4Core/Application/python/g4SimHits_cfi.py:668]:
```suggestion
phase2_common.toModify( g4SimHits, LHCTransport = False ) 
```

**Review by perrotta — COMMENTED** [review]
Sorry, when I wrote my suggestions I did not take care of the capitalizations.
Since you have to fix anyhow, there are a few more modifiers in that python file that could be also adjusted by removing the type specifications:
```
--- a/SimG4Core/Application/python/g4SimHits_cfi.py
+++ b/SimG4Core/Application/python/g4SimHits_cfi.py
@@ -629,28 +629,28 @@ run2_HCAL_2017.toModify( g4SimHits, HCalSD = dict( TestNumberingScheme = True )
 ##
 from Configuration.Eras.Modifier_run3_common_cff import run3_common
 run3_common.toModify( g4SimHits, CastorSD = dict( useShowerLibrary = False ) ) 
-run3_common.toModify( g4SimHits, LHCTransport = cms.bool(True) )
+run3_common.toModify( g4SimHits, LHCTransport = True )
 
 ##
 ## Disable PPS from Run 3 PbPb runs
 ##
 from Configuration.Eras.Modifier_pp_on_PbPb_run3_cff import pp_on_PbPb_run3
-pp_on_PbPb_run3.toModify( g4SimHits, LHCTransport = false )
+pp_on_PbPb_run3.toModify( g4SimHits, LHCTransport = False )
 
 ##
 ## Change ECAL time slices
 ##
 from Configuration.Eras.Modifier_phase2_timing_cff import phase2_timing
 phase2_timing.toModify( g4SimHits.ECalSD,
-                             StoreLayerTimeSim = cms.untracked.bool(True),
-                             TimeSliceUnit = cms.double(0.001) )
+                             StoreLayerTimeSim = True,
+                             TimeSliceUnit = 0.001 )
 ##
 ## Change CALO Thresholds
 ##
 from Configuration.Eras.Modifier_h2tb_cff import h2tb
 h2tb.toModify(g4SimHits.CaloSD,
-              EminHits  = cms.vdouble(0.0,0.0,0.0,0.0,0.0),
-              TmaxHits  = cms.vdouble(1000.0,1000.0,1000.0,1000.0,2000.0) )
+              EminHits  = [0.0,0.0,0.0,0.0,0.0],
+              TmaxHits  = [1000.0,1000.0,1000.0,1000.0,2000.0] )
 
 ##
 ## DD4hep migration
@@ -665,4 +665,4 @@ dd4hep.toModify( g4SimHits, g4GeometryDD4hepSource = True )
 from Configuration.Eras.Modifier_phase2_common_cff import phase2_common
 phase2_common.toModify(g4SimHits,
                        OnlySDs = ['ZdcSensitiveDetector', 'TotemT2ScintSensitiveDetector', 'TotemSensitiveDetector', 'RomanPotSensitiveDetector', 'PLTSensitiveDetector', 'MuonSensitiveDetector', 'MtdSensitiveDetector', 'BCM1FSensitiveDetector', 'EcalSensitiveDetector', 'CTPPSSensitiveDetector', 'HGCalSensitiveDetector', 'BSCSensitiveDetector', 'CTPPSDiamondSensitiveDetector', 'FP420SensitiveDetector', 'BHMSensitiveDetector', 'HFNoseSensitiveDetector', 'HGCScintillatorSensitiveDetector', 'CastorSensitiveDetector', 'CaloTrkProcessing', 'HcalSensitiveDetector', 'TkAccumulatingSensitiveDetector'] )
-phase2_common.toModify( g4SimHits, LHCTransport = false ) 
+phase2_common.toModify( g4SimHits, LHCTransport = False ) 
```

**Comment by civanch** [comment]
please test

**Comment by civanch** [comment]
I have double checked - let us see if it is working.

**Comment by civanch** [comment]
+1

regression is lost due to use of random numbers by PPS.

**Comment by perrotta** [comment]
+1

**Pull request merged**